### PR TITLE
std: implement the `random` feature

### PR DESCRIPTION
**Pull request by joboet**
Implements the ACP https://github.com/rust-lang/libs-team/issues/393.

# API

This PR contains two minor deviations to the API proposed in the ACP (I can easily undo them if necessary):
* `gen_bytes` is renamed to `fill_bytes`
* `DefaultRng` is renamed to `DefaultRandomSource`

Most names in `std` do not contain abbreviations, and even the documentation proposed in the ACP describes `gen_bytes` with "*Fill* `buf` with random *bytes*". `DefaultRandomSource` mirrors `DefaultHasher`, both names contain the relevant trait.

# Implementation

The implementation in this PR mostly reuses the old hash-map key generation code, but contains a few deviations:
* The Linux code was amended to `poll` `/dev/random` before reading `/dev/urandom` which is required to ensure that the entropy pool is initialized.
* The fallback code to `/dev/urandom` has been removed for FreeBSD and some other UNIX-like OSs, they now all use `getentropy` instead of `getrandom`. This function is always available on all of them (FreeBSD 12 is our minimum version, see the module documentation in `sys/random/unix.rs` for the relevant documentation) and has been included in the lastest version of the POSIX standard, issue 8, so using it whereever possible seems like a good idea, especially since NetBSD even indicates that `getrandom` [might be removed in the future](https://man.netbsd.org/getrandom.2). Falling back is unnecessary as if `getentropy` fails, `/dev/urandom` will, too, as FreeBSD [does not support](https://github.com/freebsd/freebsd-src/commit/86def3dcd66af169aac0c6ff91dd94a98369cc89) retrieving insecure random data.

I'd appreciate help from all platform maintainers to help verify that all of these implementations are correct and use a suitable API.

# Discussion about the API

I have some issues with the API myself, but to keep this PR focussed on discussion the actual implementation, I'd like to ask everyone to keep these discussions to the [ACP issue](https://github.com/rust-lang/libs-team/issues/393).

### CODE REVIEW

File: library/core/src/random.rs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+/// modulo operations, some of the resulting values can become more likely than
+/// others. Use audited crates when in doubt.
+#[unstable(feature = "random", issue = "none")]
+pub trait Random: Sized {
+    /// Generates a random value.
+    fn random(source: &mut (impl RandomSource + ?Sized)) -> Self;
+}
+
+impl Random for bool {
+    fn random(source: &mut (impl RandomSource + ?Sized)) -> Self {
```
Comment (fbstj) [inline library/core/src/random.rs:27]: nit: missing comment?

File: library/core/src/random.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    ///
+    /// **Warning:** Be careful when manipulating the resulting value! This
+    /// method samples according to a uniform distribution, so a value of 1 is
+    /// just as likely as [`MAX`](Self::MAX). By using modulo operations, some
+    /// values can become more likely than others. Use audited crates when in
+    /// doubt.
+    fn random(source: &mut (impl RandomSource + ?Sized)) -> Self {
+        u8::random(source) as i8
+    }
+}
```
Comment (tdittr) [inline library/core/src/random.rs:58]: Just scrolled by here and I'm curious. Why are `u8` and `i8` not also implemented with the macro below? They also have `to_ne_bytes` and `from_ne_bytes` so might work nicely.

File: library/std/src/sys/pal/windows/pipe.rs [hunk 3]
```diff
@@ -207,17 +206,6 @@ pub fn spawn_pipe_relay(
...
-fn random_number() -> usize {
-    static N: AtomicUsize = AtomicUsize::new(0);
-    loop {
-        if N.load(Relaxed) != 0 {
-            return N.fetch_add(1, Relaxed);
-        }
-
-        N.store(hashmap_random_keys().0 as usize, Relaxed);
-    }
-}
```
Comment (ChrisDenton) [inline library/std/src/sys/pal/windows/pipe.rs:219]: Removing this entirely changes the implementation. I'd prefer to just replace `hashmap_random_keys` here and leave any consequential changes to another PR.

File: library/core/src/random.rs (hunk 1 above)
Comment (joboet) [reply library/core/src/random.rs:27]: No. I only added the comment in the integer case to hopefully stop people from doing stupid stuff like `u8::random(...) % 6`. `bool` is much harder to misuse, the comment on the trait definition is enough.

**Comment by joboet** [comment]
While implementing, I couldn't help but notice that quite a lot of platforms warn against using their randomness syscalls for non-seed-generating purposes. For instance, the newly introduced POSIX function [`getentropy`](https://pubs.opengroup.org/onlinepubs/9799919799/functions/getentropy.html) specifies that:

"The intended use of this function is to create a seed for other pseudo-random number generators."

My gut instinct is that people will attempt to use `DefaultRandomSource` for much more than that, meaning we expose platforms to loads of small random data request, which they might not be equipped to deal with. So perhaps we should use things like `arc4random_buf` instead and clarify that the random numbers are not *the* highest quality the system can provide (but good enough in practise, `arc4random` is still cryptographically secure)?

**Comment by joshtriplett** [comment]
No objection to renaming to DefaultRandomSource. I'd personally prefer not to rename gen_bytes to fill_bytes though. It's a random number *generator*, and if people do end up calling it directly I think that name would be more evocative. This isn't a blocker, though, and if others on libs-api want it renamed I won't object.

**Comment by the8472** [comment]
There are edge-cases where anything short of the OS entropy provider or cpu instructions will not be suitable for cryptographic use. If it involves `fork()` or cloning VM snapshots you end up with reusing RNG state and thus potential key/nonce reuse which can be catastrophic for some cryptographic schemes. A libc implementation can guard against forking, but not against VM clones.

> use things like arc4random_buf instead and clarify that the random numbers are not the highest quality the system can provide (but good enough in practise, arc4random is still cryptographically secure)?

Old implementations that actually use RC4 are broken. Newer ones that switch to modern stream ciphers should be secure except for the reuse issue.

**Comment by clarfonthey** [comment]
You mention discussing on the ACP, but the ACP has been approved and is closed. You should file a tracking issue and link that on the features you implement so we can better track design concerns there.

**Comment by joboet** [comment]
> I'd personally prefer not to rename gen_bytes to fill_bytes though. It's a random number _generator_, and if people do end up calling it directly I think that name would be more evocative.

How about just `generate`, then? It's short, clear and not an abbreviation. Either way, I don't really mind.

**Comment by cuviper** [comment]
Given the continued discussion on the ACP(s), I think this should have an API reviewer.

r? libs-api

File: library/std/src/random.rs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+/// A best-effort random source used for creating hash-map seeds.
+#[unstable(feature = "random", issue = "none")]
+pub(crate) struct BestEffortRandomSource;
+
+impl RandomSource for BestEffortRandomSource {
+    fn fill_bytes(&mut self, bytes: &mut [u8]) {
+        sys::fill_bytes(bytes, true)
+    }
+}
```
Comment (juntyr) [inline library/std/src/random.rs:75]: Will this type only ever be used for creating hash-map seeds?

As noted in #129402, a future wasip2 implementation might use wasi:random/insecure for best effort randomness and wasi:random/random for the default random source. However, wasip2 also has the specific wasi:random/insecure-seed interface which seems an even closer fit for generating hash map keys.

If this type would only ever be used for initialising hash map random keys, the wasip2 implementation could just use wasi:random/insecure-seed. However, if this type became more widely used inside std (or even exposed), it might need to fall back to wasi:random/insecure.

Comment (joboet) [reply library/std/src/random.rs:75]: Fair point, I'll just reintroduce `hashmap_random_seeds`.

**Comment by joboet** [comment]
Rebased to reintroduce `hashmap_random_keys`, fix a Linux `poll` bug (a timeout of zero represents immediate timeout, not infinite as I though) and use allocation addresses for weak random data on the unsupported platforms.

**Comment by joboet** [comment]
Closing in favour of #129201.
